database.go

```diff
@@ -149,12 +163,104 @@ func (db *Database) AddBinary(binaryContent []byte) *Binary {
 	return db.Content.Meta.Binaries.Add(binaryContent, WithKDBXv31Binary)
 }
 
+type BinarySet map[*Binary]bool
+type BinaryDuplicates map[*Binary]*Binary
+type BinaryUsages map[int][]BinaryReference
+
+func (db *Database) CleanupBinaries() int {
+	binaries := Binaries{}
+	usages := db.getBinariesMap()
+	garbage := db.FindBinaryGarbage(usages)
+	duplicates := db.FindBinaryDuplicates()
+
+	for _, binary := range *db.GetBinaries() {
+		if _, ok := garbage[&binary]; ok {
+			continue
+		}
+		if original, ok := duplicates[&binary]; ok {
+			for _, ref := range usages[binary.ID] {
+				ref.Value.ID = original.ID
+			}
+			continue
+		}
+		binaries = append(binaries, binary)
+	}
+
+	counter := 0
+	for _, binary := range binaries {
+		for _, ref := range usages[binary.ID] {
+			binary.ID = counter
+			ref.Value.ID = binary.ID
+		}
+		counter++
+	}
+
+	*db.GetBinaries() = binaries
+	return len(garbage) + len(duplicates)
+}
+
+func (db *Database) FindBinaryDuplicates() BinaryDuplicates {
```

Comment:
I don't think this method needs to be exported. It should always just be called in the context of a cleanup or when adding a new file

Reply:
I think this method should be built to be called on `AddBinary`. If an identical attachment is already there, it should be skipped and just a new binary reference should be returned. That could also make the cleanup process simpler.

Any files that already have files attached multiple times might have done so intentionally, but I don't consider it to be the responsibility of gokeepasslib to also take care of that.

Reply:
Agree, but it will not work if database already contains duplicates. Hope it will never happens, going to implement in your way :)

Reply:
Then it will keep containing duplicates. It takes up a bit more storage, but that is ok as well then.